Mathlib/Computability/NFA.lean

```diff
@@ -115,6 +123,217 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+    languages. -/
+
+section Reversal
+
+section Auxilary
+
+/-- `M.unstep` is the reverse of `M.step`. -/
+def unstep (s : σ) (a : α) : Set σ := {s' | s ∈ M.step s' a}
+
+theorem mem_unstep {s t : σ} {a : α} : s ∈ M.unstep t a ↔ t ∈ M.step s a := Iff.rfl
+
+/-- Reversed analog of `M.stepSet S a`:
+  `M.unstepSet S a` is the union of `unstep M s a` for all `s ∈ S`.
+  It computes the set of states that have a transition in `M`
+  to a state in `S`. -/
+def unstepSet (S : Set σ) (a : α) : Set σ := ⋃ s ∈ S, M.unstep s a
+
+theorem mem_unstepSet {s : σ} {S : Set σ} {a : α} :
+    s ∈ M.unstepSet S a ↔ ∃ t ∈ S, s ∈ M.unstep t a := by
+  simp [unstepSet]
+
+theorem mem_unstepSet_step {s : σ} {S : Set σ} {a : α} :
+    s ∈ M.unstepSet S a ↔ ∃ t ∈ S, t ∈ M.step s a := by
+  simp [mem_unstepSet, mem_unstep]
+
+/-- Reseversed analog of `M.evalFrom S x`:
+  `M.rewindFrom S x` computes all possible reversed paths through `M` with
+  input `x` starting at an element of `S`. -/
+def rewindFrom : Set σ → List α → Set σ := List.foldl M.unstepSet
+
+/-- `M.rewind x` computes all possible paths through `M` with input `x` ending at an element of
+  `M.accept`. -/
+def rewind : List α → Set σ := M.rewindFrom M.accept
+
+/-- `M.rewindsToStart S` is the language of `x`
+  such that starting from `S` we rewind to `M.start`. -/
+def rewindsToStart (S : Set σ) : Language α :=
+  { x | ∃ s ∈ M.start, s ∈ M.rewindFrom S x }
+
+lemma mem_rewindsToStart {S : Set σ} {x : List α} :
+    x ∈ M.rewindsToStart S ↔ ∃ s ∈ M.start, s ∈ M.rewindFrom S x := Iff.rfl
+
+end Auxilary
+
+/-- NFAs are closed under reversal:
+  Given NFA `M`, there is an NFA `reverse(M)` such that
+  `L(reverse(M)) = reverse(L(M))`. -/
+def reverse (M : NFA α σ) : NFA α σ where
+  step := M.unstep
+  start := M.accept
+  accept := M.start
+
+lemma reverse_acceptsFrom_rewindsToStart {M : NFA α σ} :
+    M.reverse.acceptsFrom = M.rewindsToStart := by
+  ext xs
+  rfl
+
+lemma reverse_rewindFrom_evalFrom {xs : List α} {S1 S2 : Set σ} {M : NFA α σ} :
+    (∃ s1 ∈ S1, s1 ∈ M.rewindFrom S2 xs) ↔
+    (∃ s2 ∈ S2, s2 ∈ M.evalFrom S1 xs.reverse) := by
+  dsimp [evalFrom, rewindFrom]
+  rw [List.foldl_reverse]
+  induction xs generalizing S1 S2
+  case nil => tauto
+  case cons x xs ih =>
+    simp [ih, unstepSet, stepSet, Set.mem_iUnion]
+    tauto
+
+lemma reverse_rewindsToStart_acceptsFrom {xs : List α} {M : NFA α σ} :
+    xs ∈ M.rewindsToStart M.reverse.start ↔ xs.reverse ∈ M.acceptsFrom M.start := by
+  rw [mem_rewindsToStart, mem_acceptsFrom]
+  apply reverse_rewindFrom_evalFrom
+
+theorem reverse_accepts {M : NFA α σ} :
+    M.reverse.accepts = { xs : List α | xs.reverse ∈ M.accepts } := by
+  ext xs
+  rw [accepts_acceptsFrom, accepts_acceptsFrom, reverse_acceptsFrom_rewindsToStart, Set.mem_setOf]
+  apply reverse_rewindsToStart_acceptsFrom
+
+end Reversal
+
+section Union
+
+variable {σ1 : Type v} {σ2 : Type v}
+
+private instance Language.instUnion : Union (Language α) := by
+  apply Set.instUnion
```

Comment:
Why are you putting an instance about `Language` here? Also, it's private, but you're using it to state the main theorems about what `NFA.union` does, so... how is anyone ever supposed to make use of `NFA.union`?

Also, notation for the union of languages already exists, and it's `l₁ + l₂`.

Reply:
My apologies, I did not catch that. I replaced this instance with the `Add` instance for `Language`, and I moved the `Inter` instance I created to `Language.lean`.